request.go

```diff
@@ -0,0 +1,184 @@
+	// to the request in the future. OPTIONAL.
+	Extensions []pkix.Extension
+}
+
+// NewRequest creates a timestamp request based on caller provided options.
+func NewRequest(opts RequestOptions) (*Request, error) {
+	if opts.Content == nil {
+		return nil, &MalformedRequestError{Msg: "content to be time stamped cannot be empty"}
+	}
+	hashAlg, err := oid.FromHash(opts.HashAlgorithm)
+	if err != nil {
+		return nil, &MalformedRequestError{Msg: err.Error()}
+	}
+	digest, err := hashutil.ComputeHash(opts.HashAlgorithm, opts.Content)
+	if err != nil {
+		return nil, &MalformedRequestError{Msg: err.Error()}
+	}
+	var nonce *big.Int
+	if !opts.NoNonce {
+		if opts.Nonce != nil { // user provided Nonce, use it
+			nonce = opts.Nonce
+		} else { // user ignored Nonce, use built-in Nonce
+			var err error
+			nonce, err = generateNonce()
+			if err != nil {
+				return nil, &MalformedRequestError{Msg: err.Error()}
+			}
+		}
+	}
+	return &Request{
+		Version: 1,
+		MessageImprint: MessageImprint{
+			HashAlgorithm: pkix.AlgorithmIdentifier{
+				Algorithm:  hashAlg,
+				Parameters: opts.HashAlgorithmParameters,
+			},
+			HashedMessage: digest,
+		},
+		ReqPolicy:  opts.ReqPolicy,
+		Nonce:      nonce,
+		CertReq:    !opts.NoCert,
+		Extensions: opts.Extensions,
+	}, nil
+}
+
+// MarshalBinary encodes the request to binary form.
+// This method implements encoding.BinaryMarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryMarshaler
+func (r *Request) MarshalBinary() ([]byte, error) {
+	if r == nil {
+		return nil, errors.New("nil request")
+	}
+	return asn1.Marshal(*r)
+}
+
+// UnmarshalBinary decodes the request from binary form.
+// This method implements encoding.BinaryUnmarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryUnmarshaler
+func (r *Request) UnmarshalBinary(data []byte) error {
+	_, err := asn1.Unmarshal(data, r)
+	return err
+}
+
+// Validate checks if req is a valid request against RFC 3161.
+// It is used before a timstamp requestor sending the request to TSA.
+func (req *Request) Validate() error {
+	if req == nil {
+		return &MalformedRequestError{Msg: "request cannot be nil"}
+	}
+	if req.Version != 1 {
+		return &MalformedRequestError{Msg: fmt.Sprintf("request version must be 1, but got %d", req.Version)}
+	}
+	hashAlg := req.MessageImprint.HashAlgorithm.Algorithm
+	hash, available := oid.ToHash(hashAlg)
+	if !available {
+		return &MalformedRequestError{Msg: fmt.Sprintf("hash algorithm %v is unavailable", hashAlg)}
+	}
+	if hash.Size() != len(req.MessageImprint.HashedMessage) {
+		return &MalformedRequestError{Msg: fmt.Sprintf("hashed message is of incorrect size %d", len(req.MessageImprint.HashedMessage))}
+	}
+	return nil
+}
+
+// generateNonce generates a built-in Nonce for TSA request
+func generateNonce() (*big.Int, error) {
+	// Pick a random number from 0 to 2^159
+	nonce, err := rand.Int(rand.Reader, (&big.Int{}).Exp(big.NewInt(2), big.NewInt(159), nil))
```

Comment:
Instead of using the expensive `Exp`, you may think about binary operations like `SetBit()` or `Lsh()`.

Reply:
Updated to use `Lsh()` instead.